### PR TITLE
cors: initial implementation

### PR DESCRIPTION
**Pull request by Moesang**

### CODE REVIEW

File: cors.go [hunk 1]
```diff
@@ -1 +1,150 @@
...
+		}
+		if len(opt.Scheme) == 0 {
+			opt.Scheme = "http"
+		}
+		if len(opt.AllowDomain) == 0 {
+			opt.AllowDomain = []string{"*"}
+		}
+		if !opt.AllowSubdomain {
+			opt.AllowSubdomain = false
+		}
```
Comment (wuhan005) [inline cors.go:56]:
```suggestion
```

File: cors.go [hunk 2]
```diff
@@ -1 +1,150 @@
...
+				http.MethodDelete,
+				http.MethodOptions,
+			}
+		}
+		if opt.MaxAgeSeconds <= 0 {
+			opt.MaxAgeSeconds = 600
+		}
+		if !opt.AllowCredentials {
+			opt.AllowCredentials = true
+		}
```
Comment (wuhan005) [inline cors.go:73]:
```suggestion
```

File: cors.go [hunk 3]
```diff
@@ -1 +1,150 @@
...
+	Methods          []string
+	MaxAgeSeconds    int
+	AllowCredentials bool
+}
+
+func prepareOptions(options []Options) Options {
+	var opt Options
+	if len(options) > 0 {
+		opt = options[0]
+	} else {
```
Comment (wuhan005) [inline cors.go:44]:
```suggestion
	}
```

File: cors.go [hunk 4]
```diff
@@ -1 +1,150 @@
...
+				http.MethodOptions,
+			}
+		}
+		if opt.MaxAgeSeconds <= 0 {
+			opt.MaxAgeSeconds = 600
+		}
+		if !opt.AllowCredentials {
+			opt.AllowCredentials = true
+		}
+	}
```
Comment (wuhan005) [inline cors.go:74]:
```suggestion
```

File: cors.go [hunk 5]
```diff
@@ -1 +1,150 @@
...
+				return
+			}
+
+			u, err := url.Parse(origin)
+			if err != nil {
+				respErrorf(ctx, log, http.StatusBadRequest, "Failed to parse CORS origin header. Reason: %v", err)
+				return
+			}
+
+			ok := false
```
Comment (wuhan005) [inline cors.go:107]:
```suggestion
			var ok bool
```

File: cors.go [hunk 6]
```diff
@@ -1 +1,150 @@
...
+		}
+
+		ctx.ResponseWriter().Before(func(w flamego.ResponseWriter) {
+			for k, v := range headers {
+				w.Header().Set(k, v)
+			}
+		})
+
+		if reqOptions {
+			ctx.ResponseWriter().WriteHeader(200) // return response
```
Comment (wuhan005) [inline cors.go:135]:
```suggestion
			ctx.ResponseWriter().WriteHeader(http.StatusOK)
```

File: cors_test.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+func TestCORS(t *testing.T) {
+	f := flamego.NewWithLogger(&bytes.Buffer{})
+	f.Use(CORS())
+	f.Get("/", func(c flamego.Context) string {
+		return "ok"
+	})
+
+	resp := httptest.NewRecorder()
+	req, err := http.NewRequest(http.MethodGet, "/", strings.NewReader("some request"))
```
Comment (wuhan005) [inline cors_test.go:22]:
```suggestion
	req, err := http.NewRequest(http.MethodGet, "/", nil)
```

**Review by wuhan005 — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -7,6 +7,24 @@
...
+	"github.com/flamego/cors"
+)
+func main() {
+	f := flamego.Classic()
+	f.Use(cors.CORS())
+	f.Get("/", func(c flamego.Context) string {
+        return "ok"
+    })
+	f.Run()
+}
```
Comment (unknwon) [inline README.md:25]: CleanShot 2021-07-31 at 22 20 33@2x

Mixed use of tabs and spaces.

File: cors.go [hunk 9]
```diff
@@ -1 +1,144 @@
...
+Special values are the  a single '*' wildcard that will allow any domain to send requests without
+credentials and the special '!*' wildcard which will reply with requesting domain in the 'access-control-allow-origin'
+header and hence allow requests from any domain *with* credentials.
+ALLOW_SUBDOMAIN set to true accepts requests from any subdomains of ALLOW_DOMAIN.
+METHODS may be a comma separated list of HTTP-methods to be accepted.
+MAX_AGE_SECONDS may be the duration in secs for which the response is cached (default 600).
+ref: https://stackoverflow.com/questions/54300997/is-it-possible-to-cache-http-options-response?noredirect=1#comment95790277_54300997
+ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+ALLOW_CREDENTIALS set to false rejects any request with credentials.
+*/
```
Comment (unknwon) [inline cors.go:29]:
```suggestion
/*
Options to configure the CORS middleware read from the [cors] section of the ini configuration file.
SCHEME may be http or https as accepted schemes or the '*' wildcard to accept any scheme.
ALLOW_DOMAIN may be a comma separated list of domains that are allowed to run CORS requests
Special values are the  a single '*' wildcard that will allow any domain to send requests without
credentials and the special '!*' wildcard which will reply with requesting domain in the 'access-control-allow-origin'
header and hence allow requests from any domain *with* credentials.
ALLOW_SUBDOMAIN set to true accepts requests from any subdomains of ALLOW_DOMAIN.
METHODS may be a comma separated list of HTTP-methods to be accepted.
MAX_AGE_SECONDS may be the duration in secs for which the response is cached (default 600).
ref: https://stackoverflow.com/questions/54300997/is-it-possible-to-cache-http-options-response?noredirect=1#comment95790277_54300997
ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
ALLOW_CREDENTIALS set to false rejects any request with credentials.
*/
```

Use `//` for comments almost always.

File: README.md [hunk 10]
```diff
@@ -7,6 +7,24 @@
...
+    "github.com/flamego/cors"
+)
+func main() {
+    f := flamego.Classic()
+    f.Use(cors.CORS())
+    f.Get("/", func(c flamego.Context) string {
+        return "ok"
+    })
+    f.Run()
+}
```
Comment (unknwon) [inline README.md:25]:
```suggestion
package main

import (
	"github.com/flamego/flamego"
	"github.com/flamego/cors"
)

func main() {
	f := flamego.Classic()
	f.Use(cors.CORS())
	f.Get("/", func(c flamego.Context) string {
		return "ok"
	})
	f.Run()
}
```

File: cors.go [hunk 11]
```diff
@@ -1 +1,143 @@
...
+// Special values are the  a single '*' wildcard that will allow any domain to send requests without
+// credentials and the special '!*' wildcard which will reply with requesting domain in the 'access-control-allow-origin'
+// header and hence allow requests from any domain *with* credentials.
+// ALLOW_SUBDOMAIN set to true accepts requests from any subdomains of ALLOW_DOMAIN.
+// METHODS may be a comma separated list of HTTP-methods to be accepted.
+// MAX_AGE_SECONDS may be the duration in secs for which the response is cached (default 600).
+// ref: https://stackoverflow.com/questions/54300997/is-it-possible-to-cache-http-options-response?noredirect=1#comment95790277_54300997
+// ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+// ALLOW_CREDENTIALS set to false rejects any request with credentials.
+
```
Comment (unknwon) [inline cors.go:28]:
```suggestion
```

File: cors.go [hunk 12]
```diff
@@ -1 +1,142 @@
...
+// ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+// ALLOW_CREDENTIALS set to false rejects any request with credentials.
+type Options struct {
+	Section          string
+	Scheme           string
+	AllowDomain      []string
+	AllowSubdomain   bool
+	Methods          []string
+	MaxAgeSeconds    int
+	AllowCredentials bool
```
Comment (unknwon) [inline cors.go:35]: We need docstring for every public field.

File: cors.go [hunk 13]
```diff
@@ -1 +1,142 @@
...
+// CORS responds to preflight requests with adequate access-control-* respond headers
+// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+// https://fetch.spec.whatwg.org/#cors-protocol-and-credentials
+func CORS(options ...Options) flamego.Handler {
+	opt := prepareOptions(options)
+	return func(ctx flamego.Context, log *log.Logger) {
+		reqOptions := ctx.Request().Method == http.MethodOptions
+
+		headers := map[string]string{
+			"access-control-allow-methods": strings.Join(opt.Methods, ","),
```
Comment (unknwon) [inline cors.go:80]:
```suggestion
			"Access-Control-Allow-Methods": strings.Join(opt.Methods, ","),
```

File: cors.go [hunk 14]
```diff
@@ -1 +1,142 @@
...
+func respErrorf(ctx flamego.Context, log *log.Logger, statusCode int, format string, a ...interface{}) {
+	msg := fmt.Sprintf(format, a...)
+	log.Println(msg)
+	ctx.ResponseWriter().WriteHeader(statusCode)
+
+	_, err := ctx.ResponseWriter().Write([]byte(msg))
+	if err != nil {
+		panic(err)
+	}
+}
```
Comment (unknwon) [inline cors.go:142]: Why can't we simply use `http.Error` and remove this function?

File: README.md [hunk 15]
```diff
@@ -7,6 +7,26 @@
...
 Package cors is a middleware that provides the Cross-Origin Resource Sharing.
 
+## Getting started
+
+```go
+package main
+
+import (
+	"github.com/flamego/flamego"
+	"github.com/flamego/cors"
```
Comment (unknwon) [inline README.md:17]:
```suggestion
	"github.com/flamego/cors"
	"github.com/flamego/flamego"
```

`c` is before `f`, and I think this is what `gofmt` would resort for us.

File: cors.go [hunk 16]
```diff
@@ -1 +1,132 @@
...
+	"strings"
+
+	"github.com/flamego/flamego"
+)
+
+const anyDomain = "!*"
+
+// Options to configure the CORS middleware read from the [cors] section of the ini configuration file.
+// ref: https://stackoverflow.com/questions/54300997/is-it-possible-to-cache-http-options-response?noredirect=1#comment95790277_54300997
+// ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
```
Comment (unknwon) [inline cors.go:23]:
```suggestion
// Options contains options for the corS.CORS middleware.
```

File: cors.go [hunk 17]
```diff
@@ -1 +1,132 @@
...
+	// header and hence allow requests from any domain *with* credentials.
+	AllowDomain []string
+	// AllowSubdomain allowed
+	AllowSubdomain bool
+	// METHODS may be a comma separated list of HTTP-methods to be accepted.
+	Methods []string
+	// MAX_AGE_SECONDS may be the duration in secs for which the response is cached (default 600).
+	MaxAgeSeconds int
+	// ALLOW_CREDENTIALS set to false rejects any request with credentials.
+	AllowCredentials bool
```
Comment (unknwon) [inline cors.go:39]:
```suggestion
	// Scheme may be http or https as accepted schemes or the '*' wildcard to accept any scheme.
	Scheme string
	// AllowDomains is a comma separated list of domains that are allowed to run CORS requests
	// Special values are the  a single '*' wildcard that will allow any domain to send requests without
	// credentials and the special '!*' wildcard which will reply with requesting domain in the 'access-control-allow-origin'
	// header and hence allow requests from any domain *with* credentials.
	AllowDomains []string
	// AllowSubdomain allowed ???
	AllowSubdomain bool
	// Methods is a comma separated list of HTTP-methods to be accepted.
	Methods []string
	// MaxAgeSeconds is the duration in secs for which the response is cached (default 600).
	MaxAgeSeconds int
	// AllowCredentials set to false rejects any request with credentials.
	AllowCredentials bool
```

1. Please finish the incomplete sentence.
2. Please describe what are the default values for each field if left unset.

File: cors.go [hunk 18]
```diff
@@ -1 +1,132 @@
...
+	}
+	if len(opt.Methods) == 0 {
+		opt.Methods = []string{
+			http.MethodGet,
+			http.MethodHead,
+			http.MethodPost,
+			http.MethodPut,
+			http.MethodPatch,
+			http.MethodDelete,
+			http.MethodOptions,
```
Comment (unknwon) [inline cors.go:62]: Why is all of these methods necessary as the default list?

**Review by unknwon — CHANGES_REQUESTED** [review]

File: cors.go (hunk 18 above)
Comment (Moesang) [reply cors.go]: The list has remaind only GET, POST and OPTIONS now.

File: cors.go [hunk 19]
```diff
@@ -1 +1,126 @@
...
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/flamego/flamego"
+)
+
+const anyDomain = "!*"
```
Comment (unknwon) [inline cors.go:19]: Why does this constant needs to be defined as global?

File: cors.go [hunk 20]
```diff
@@ -1 +1,126 @@
...
+
+// Options contains options for the cors.CORS middleware.
+type Options struct {
+	// Scheme may be http or https as accepted schemes or the '*' wildcard to accept any scheme. (default: "http")
+	Scheme string
```
Comment (unknwon) [inline cors.go:24]:
```suggestion
	// Scheme may be http or https as accepted schemes or the '*' wildcard to accept any scheme. Default is "http".
	Scheme string
```

Please take a look at how field docstring is written in other Options:

https://github.com/flamego/flamego/blob/eff1d929e6ce0aadc12c1493918d02bdb0d04798/render.go#L33-L42

Same for other fields in the struct.

File: cors.go [hunk 21]
```diff
@@ -1 +1,126 @@
...
+	// Special values are the  a single '*' wildcard that will allow any domain to send requests without
+	// credentials and the special '!*' wildcard which will reply with requesting domain in the 'access-control-allow-origin'
+	// header and hence allow requests from any domain *with* credentials. (default '*')
+	AllowDomain []string
+	// AllowSubdomain allowed subdomains of domains to run CORS requests. (default false)
+	AllowSubdomain bool
+	// Methods may be a comma separated list of HTTP-methods to be accepted. (default GET, POST, OPTIONS)
+	Methods []string
+	// MaxAgeSeconds may be the duration in secs for which the response is cached. (default 600)
+	MaxAgeSeconds int
```
Comment (unknwon) [inline cors.go:35]:
```suggestion
	MaxAge time.Duration
```

I think this is much easier to use in Go, and when we use it, we just call `MaxAge.Seconds()`

File: cors.go [hunk 22]
```diff
@@ -1 +1,126 @@
...
+	AllowCredentials bool
+}
+
+func prepareOptions(options []Options) Options {
+	var opt Options
+	if len(options) > 0 {
+		opt = options[0]
+	}
+
+	if len(opt.Scheme) == 0 {
```
Comment (unknwon) [inline cors.go:46]:
```suggestion
	if opt.Scheme == "" {
```

The idiomatic Go is to compare with empty string when checking if the string is empty.

File: cors.go [hunk 23]
```diff
@@ -1 +1,126 @@
...
+	if opt.MaxAgeSeconds <= 0 {
+		opt.MaxAgeSeconds = 600
+	}
+
+	return opt
+}
+
+// CORS responds to preflight requests with adequate access-control-* respond headers
+// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+// https://fetch.spec.whatwg.org/#cors-protocol-and-credentials
```
Comment (unknwon) [inline cors.go:68]:
```suggestion
// CORS returns a middleware handler that responds to preflight requests with adequate "Access-Control-*" response headers.
```

File: cors.go [hunk 24]
```diff
@@ -1 +1,126 @@
...
+func CORS(options ...Options) flamego.Handler {
+	opt := prepareOptions(options)
+	return func(ctx flamego.Context, log *log.Logger) {
```
Comment (unknwon) [inline cors.go:71]: Seem we're not using `log *log.Logger` at all, can we remove it?

File: cors.go [hunk 25]
```diff
@@ -1 +1,126 @@
...
+		}
+
+		ctx.ResponseWriter().Before(func(w flamego.ResponseWriter) {
+			for k, v := range headers {
+				w.Header().Set(k, v)
+			}
+		})
+
+		if reqOptions {
+			ctx.ResponseWriter().WriteHeader(http.StatusOK) // return response
```
Comment (unknwon) [inline cors.go:122]: Why do we only write response code for OPTIONS method? What about other methods (e.g. GET, POST)?

**Review by unknwon — CHANGES_REQUESTED** [review]

File: cors.go (hunk 25 above)
Comment (Moesang) [reply cors.go:122]: A CORS-preflight request method is OPTIONS, it seems not required to response other methods. 
reference: https://fetch.spec.whatwg.org/#cors-preflight-fetch

Comment (unknwon) [reply cors.go:127]: If that's the case, why would we want to continue if it's not a OPTIONS request? Can't we just return here?

File: cors.go [hunk 26]
```diff
@@ -1 +1,130 @@
...
+				if opt.Scheme != "*" {
+					u.Scheme = opt.Scheme
+				}
+				headers["Access-Control-Allow-Origin"] = u.String()
+				headers["Access-Control-Allow-Credentials"] = strconv.FormatBool(opt.AllowCredentials)
+				headers["Vary"] = "Origin"
+			} else {
+				http.Error(ctx.ResponseWriter(), fmt.Sprintf("CORS request from prohibited domain %v", origin), http.StatusBadRequest)
+				return
+			}
```
Comment (unknwon) [inline cors.go:119]:
```suggestion
			if !ok {
				http.Error(ctx.ResponseWriter(), fmt.Sprintf("CORS request from prohibited domain %v", origin), http.StatusBadRequest)
				return
			}

			if opt.Scheme != "*" {
				u.Scheme = opt.Scheme
			}
			headers["Access-Control-Allow-Origin"] = u.String()
			headers["Access-Control-Allow-Credentials"] = strconv.FormatBool(opt.AllowCredentials)
			headers["Vary"] = "Origin"
```

File: cors_test.go [hunk 27]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2021 Flamego. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Package cors is a middleware that generates CORS headers for Flamego.
```
Comment (unknwon) [inline cors_test.go:5]:
```suggestion
```

File: cors_test.go [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+package cors
+
+import (
+	"bytes"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/flamego/flamego"
+	"github.com/stretchr/testify/assert"
```
Comment (unknwon) [inline cors_test.go:15]:
```suggestion
	"github.com/stretchr/testify/assert"

	"github.com/flamego/flamego"
```

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the followup!

Left some more comments, and looking at tests, I'm certain many cases are not tested, please take some thoughts to improve :)

File: cors_test.go [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+
+			f.ServeHTTP(resp, req)
+
+			for headerKey, headerValue := range test.respHeaders {
+				assert.Equal(t, headerValue, resp.Header().Get(headerKey))
+			}
+		})
+	}
+
+	f2 := flamego.NewWithLogger(&bytes.Buffer{})
```
Comment (unknwon) [inline cors_test.go:63]: Why not put this set of tests in a separate function instead of naming it `f2`?

File: cors_test.go [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+		return "ok"
+	})
+
+	defaultTests := []struct {
+		name        string
+		method      string
+		respHeaders map[string]string
+	}{
+		{
+			name:   "Error method",
```
Comment (unknwon) [inline cors_test.go:32]:
```suggestion
			name:   "error method",
```

File: cors_test.go [hunk 31]
```diff
@@ -0,0 +1,164 @@
...
+			method: http.MethodGet,
+			respHeaders: map[string]string{
+				"Access-Control-Allow-Origin": "",
+				"Access-Control-Max-Age":      "",
+			},
+		},
+		{
+			name:   "Default cors response",
```
Comment (unknwon) [inline cors_test.go:40]:
```suggestion
			name:   "default response",
```

File: cors_test.go [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+func TestCORS(t *testing.T) {
+	f := flamego.NewWithLogger(&bytes.Buffer{})
+	f.Use(CORS())
+	f.Get("/", func(c flamego.Context) string {
+		return "ok"
+	})
+
+	defaultTests := []struct {
```
Comment (unknwon) [inline cors_test.go:26]:
```suggestion
	tests := []struct {
```

File: cors_test.go [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+	for _, test := range defaultTests {
+		t.Run(test.name, func(t *testing.T) {
+			resp := httptest.NewRecorder()
+			req, err := http.NewRequest(test.method, "/", nil)
+			assert.Nil(t, err)
+
+			f.ServeHTTP(resp, req)
+
+			for headerKey, headerValue := range test.respHeaders {
+				assert.Equal(t, headerValue, resp.Header().Get(headerKey))
```
Comment (unknwon) [inline cors_test.go:58]: Is `assert.Equal` not working for comparing maps?

File: cors_test.go [hunk 34]
```diff
@@ -0,0 +1,164 @@
...
+	f := flamego.NewWithLogger(&bytes.Buffer{})
+	f.Use(CORS())
+	f.Get("/", func(c flamego.Context) string {
+		return "ok"
+	})
+
+	defaultTests := []struct {
+		name        string
+		method      string
+		respHeaders map[string]string
```
Comment (unknwon) [inline cors_test.go:29]:
```suggestion
		wantHeaders map[string]string
```

**Review by unknwon — CHANGES_REQUESTED** [review]

File: cors_test.go [hunk 35]
```diff
@@ -0,0 +1,164 @@
...
+			method: http.MethodGet,
+			reqHeaders: map[string]string{
+				"Origin": "https://example.com",
+			},
+			respHeaders: map[string]string{
+				"Access-Control-Allow-Origin":      "",
+				"Access-Control-Max-Age":           "",
+				"Access-Control-Allow-Credentials": "",
+			},
+			statueCode: 200,
```
Comment (unknwon) [inline cors_test.go:100]:
```suggestion
			statueCode: http.StatusOK,
```

File: cors_test.go [hunk 36]
```diff
@@ -0,0 +1,165 @@
...
+	f.Get("/", func(c flamego.Context) string {
+		return "ok"
+	})
+
+	tests := []struct {
+		name        string
+		method      string
+		reqHeaders  map[string]string
+		wantHeaders map[string]string
+		statueCode  int
```
Comment (unknwon) [inline cors_test.go:89]:
```suggestion
		wantCode  int
```

**Review by unknwon — APPROVED** [review]
One last minor comment, approve to unblock.

File: cors.go [hunk 37]
```diff
@@ -1 +1,128 @@
...
+	opt := prepareOptions(options)
+	return func(ctx flamego.Context) {
+		if ctx.Request().Method != http.MethodOptions {
+			return
+		}
+
+		headers := map[string]string{
+			"Access-Control-Allow-Methods": strings.Join(opt.Methods, ","),
+			"Access-Control-Allow-Headers": ctx.Request().Header.Get("Access-Control-Request-Headers"),
+			"Access-Control-Max-Age":       fmt.Sprintf("%.0f", opt.MaxAge.Seconds()),
```
Comment (wuhan005) [inline cors.go:84]:
```go
fmt.Printf("%.0f", 123.456)
```
You will get `1230` here,

Why not simply use `strconv.Itoa(int(a.Seconds()))` instead?

**Review by wuhan005 — CHANGES_REQUESTED** [review]

File: cors.go (hunk 37 above)
Comment (wuhan005) [reply cors.go:84]: Or a better way to convert `float64` to `int`:
```
strconv.FormatFloat(123.4567890, 'f', 0, 64)// 123
```

FYI: https://pkg.go.dev/strconv#FormatFloat

Comment (Moesang) [reply cors.go]:
```
fmt.Printf("%.0f", 123.456)
```
Out put is `123` not `1230`.

Comment (wuhan005) [reply cors.go]: Oops, my mistake.
I got some messy output in my Go playground. 😓

**Review by wuhan005 — APPROVED** [review]

**Pull request merged**